crates/state-aggregate/src/lib.rs

```diff
@@ -0,0 +1,85 @@
+use serde::{Deserialize, Serialize};
+
+use flat_serialize_macro::FlatSerializable;
+
+// Intermediate state kept in postgres.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+pub struct TransState {
+    records: Vec<Record>,
+}
+
+impl TransState {
+    pub fn new() -> Self {
+        Self { records: vec![] }
+    }
+
+    pub fn record(&mut self, state: String, time: i64) {
+        self.records.push(Record { state, time });
+    }
+
+    pub fn append(&mut self, other: &mut Self) {
+        self.records.append(&mut other.records)
+    }
+
+    /// Drain accumulated state, sort, and return map of states to durations.
+    pub fn to_duration_map(&mut self) -> std::collections::BTreeMap<String, i64> {
+        self.records.sort_by(|a, b| a.time.cmp(&b.time));
+        let mut duration_state = DurationState {
+            last_state: None,
+            durations: std::collections::BTreeMap::new(),
+        };
+        for record in self.records.drain(..) {
+            duration_state.handle_record(record.state, record.time);
+        }
+        duration_state.durations
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Eq, FlatSerializable, PartialEq, Serialize)]
+#[repr(C)]
+pub struct DurationInState {
+    pub duration: i64,
+    // TODO String not FlatSerializable... what do?
+    //state: String,
+    pub state: u64,
+}
+
+struct DurationState {
+    last_state: Option<(String, i64)>,
+    // TODO using BTreeMap so lazy prototype tests can depend on stable iteration order
+    durations: std::collections::BTreeMap<String, i64>,
+}
+impl DurationState {
+    fn handle_record(&mut self, state: String, raw_time: i64) {
+        if let Some((last_state, last_time)) = self.last_state.take() {
+            if last_state == state {
+                // Same state recorded twice:  preserve the older timestamp.
+                self.last_state = Some((state, last_time));
```

Comment:
What if this is the last record?  It looks like if the last N entries have the same state, that interval won't be reflected in the result of `to_duration_map`

Reply:
I added a test for this, then fixed it, and then realized the bug was only caused by unnecessary complexity here anyway. I simply deleted the whole part about checking for two states in a row and all tests pass including the new one covering this case.